src/enable_variable_ops/ngraph_encapsulate_op.cc

```diff
@@ -0,0 +1,897 @@
+    //
+
+    // Create the vector.
+    int32 max_arg_index = -1;
+    std::vector<const Node*> arg_nodes;
+
+    for (auto node : m_graph.nodes()) {
+      if (node->type_string() == "_Arg") {
+        arg_nodes.push_back(node);
+
+        int32 index;
+        OP_REQUIRES_OK(ctx, GetNodeAttr(node->attrs(), "index", &index));
+        if (index > max_arg_index) max_arg_index = index;
+      }
+    }
+
+    m_input_is_static = std::vector<bool>(max_arg_index + 1, false);
+
+    // Fill the vector.
+    for (auto node : arg_nodes) {
+      int32 index;
+      OP_REQUIRES_OK(ctx, GetNodeAttr(node->attrs(), "index", &index));
+
+      bool is_static = false;
+      for (auto edge : node->out_edges()) {
+        if (edge->IsControlEdge() || !edge->dst()->IsOp()) {
+          continue;
+        }
+
+        NGRAPH_VLOG(5) << "For arg " << index << " checking edge "
+                       << edge->DebugString();
+
+        if (InputIsStatic(edge->dst(), edge->dst_input())) {
+          NGRAPH_VLOG(5) << "Marking edge static: " << edge->DebugString();
+          is_static = true;
+          break;
+        }
+      }
+
+      NGRAPH_VLOG(5) << "Marking arg " << index << " is_static: " << is_static;
+      m_input_is_static[index] = is_static;
+    }
+
+    // Set the backend type for the op
+    OP_REQUIRES_OK(ctx,
+                   ctx->GetAttr<string>("_ngraph_backend", &m_op_backend_name));
+    BackendManager::CreateBackend(m_op_backend_name);
+    event.Stop();
+    Event::WriteTrace(event);
+  }
+
+  //---------------------------------------------------------------------------
+  //  ~NGraphEncapsulateOp()
+  //---------------------------------------------------------------------------
+  ~NGraphEncapsulateOp() override {
+    std::ostringstream oss;
+    oss << "Destroy Encapsulate_" << my_instance_id << ": " << name();
+    Event event(oss.str().c_str(), name().c_str());
+
+    // If the kernel goes away, we must de-register all of its cached
+    // functions
+    // from the freshness tracker.
+    if (m_freshness_tracker != nullptr) {
+      for (auto kv : m_ng_exec_map) {
+        m_freshness_tracker->RemoveUser(kv.second);
+      }
+
+      // TODO(amprocte): We should be able to unref the tracker here, but it
+      // seems to screw things up in the C++ unit tests.
+      // m_freshness_tracker->Unref();
+
+      // TODO(malikshr) : Delete the [Could be erroreneous we dont know if this
+      // destructor is
+      // called at the very end]
+      string node_name = "_ngraph_cluster_" + to_string(m_ngraph_cluster);
+
+      for (int i = 0; i < m_number_outputs; i++) {
+        string key;
+        if (i == 0) {
+          key = to_string(m_graph_id) + node_name;
+        } else {
+          key = NGraphCatalog::CreateNodeKey(m_graph_id, node_name, i);
+        }
+        bool ref_exists = NGraphCatalog::ExistsInOutputCatalog(key);
+        if (ref_exists) {
+          NGraphCatalog::DeleteTensorFromEncapOutputCatalog(key);
```

Comment:
`DeleteTensorFromEncapOutputCatalog` calls `erase`. If a key does not exist the `erase` call will simply do nothing. So we do not need the `if (ref_exists)` here.

Also, if the purpose is to release the whole map at destruction time, then we can simply call `clear` instead of iterating over each key and `eras`ing them one by one. In which case, we can expose `ClearCatalog()` instead of `DeleteTensorFromEncapOutputCatalog(key)`

Reply:
I wonder if we need the call to `DeleteTensorFromEncapOutputCatalog` at all? When `NGraphEncapsulateOp` is destructing we can let the catalog be. I think, the catalog's static members's destructors will be called and the map will be emptied when the program is exiting, right?

Reply:
The purpose is to clear the tensors stored for this particular ng-encapsulate Op. Hence clearing it by looping over specific indices. 

I was getting a segfault without this . The segfault seemed to be coming from the destructor of ng-tensor on GPU backend.